In [22]:
import pandas as pd
import numpy as np
import json
import os
import datetime

In [3]:
os.getcwd()

'/Users/yiwen/Desktop/Fetch Rewards'

#### Receipts table

In [58]:
def date_transfer(milliseconds):
    if milliseconds is None:
        return None
    return datetime.datetime.fromtimestamp(milliseconds/1000.0).strftime("%Y-%m-%d")

In [80]:
receipts = []
with open('data/receipts.json') as f:
    for line in f:
        receipts.append(json.loads(line))

rcp_id_list = []
bonusPointsEarned_list = []
createDate_list = []
dateScanned_list = []
finishedDate_list = []
modifyDate_list = []
pointsAwardedDate_list = []
pointsEarned_list = []
purchaseDate_list = []
purchasedItemCount_list = []
barcode_list = []
description_list = []
finalPrice_list = []
itemPrice_list = []
quantityPurchased_list = []
rewardsReceiptStatus_list = []
totalSpent_list = []
userId_list = []

for i in range(len(receipts)):
    rcp = receipts[i]
    if rcp.get('rewardsReceiptItemList', None):
        for j in range(len(rcp['rewardsReceiptItemList'])):
            rcp_id = rcp['_id']['$oid']
            bonusPointsEarned = rcp.get('bonusPointsEarned', None)
            createDate = date_transfer(rcp.get('createDate').get('$date', None))
            dateScanned = date_transfer(rcp.get('dateScanned').get('$date', None))
            finishedDate = date_transfer(rcp.get('finishedDate', None).get('$date', None) if rcp.get('finishedDate', None) else None)
            modifyDate = date_transfer(rcp.get('modifyDate').get('$date', None))
            pointsAwardedDate = date_transfer(rcp.get('pointsAwardedDate').get('$date', None) if rcp.get('pointsAwardedDate', None) else None)
            pointsEarned = rcp.get('pointsEarned', None)
            purchaseDate = date_transfer(rcp.get('purchaseDate').get('$date', None) if rcp.get('purchaseDate', None) else None)
            purchasedItemCount = rcp.get('purchasedItemCount', None)
            barcode = rcp['rewardsReceiptItemList'][j].get('barcode', None)
            description = rcp['rewardsReceiptItemList'][j].get('description', None)
            finalPrice = rcp['rewardsReceiptItemList'][j].get('finalPrice', None)
            itemPrice = rcp['rewardsReceiptItemList'][j].get('itemPrice', None)
            quantityPurchased = rcp['rewardsReceiptItemList'][j].get('quantityPurchased', None)
            rewardsReceiptStatus = rcp.get('rewardsReceiptStatus', None)
            totalSpent = rcp.get('totalSpent', None)
            userId = rcp.get('userId', None)
            rcp_id_list.append(rcp_id)
            bonusPointsEarned_list.append(bonusPointsEarned)
            createDate_list.append(createDate)
            dateScanned_list.append(dateScanned)
            finishedDate_list.append(finishedDate)
            modifyDate_list.append(modifyDate)
            pointsAwardedDate_list.append(pointsAwardedDate)
            pointsEarned_list.append(pointsEarned)
            purchaseDate_list.append(purchaseDate)
            purchasedItemCount_list.append(purchasedItemCount)
            barcode_list.append(barcode)
            description_list.append(description)
            finalPrice_list.append(finalPrice)
            itemPrice_list.append(itemPrice)
            quantityPurchased_list.append(quantityPurchased)
            rewardsReceiptStatus_list.append(rewardsReceiptStatus)
            totalSpent_list.append(totalSpent)
            userId_list.append(userId) 

    else:
        rcp_id = rcp['_id']['$oid']
        bonusPointsEarned = rcp.get('bonusPointsEarned', None)
        createDate = date_transfer(rcp.get('createDate').get('$date', None))
        dateScanned = date_transfer(rcp.get('dateScanned').get('$date', None))
        finishedDate = date_transfer(rcp.get('finishedDate', None).get('$date', None) if rcp.get('finishedDate', None) else None)
        modifyDate = date_transfer(rcp.get('modifyDate').get('$date', None))
        pointsAwardedDate = date_transfer(rcp.get('pointsAwardedDate').get('$date', None) if rcp.get('pointsAwardedDate', None) else None)
        pointsEarned = rcp.get('pointsEarned', None)
        purchaseDate = date_transfer(rcp.get('purchaseDate').get('$date', None) if rcp.get('purchaseDate', None) else None)
        purchasedItemCount = rcp.get('purchasedItemCount', None)
        barcode = None
        description = None
        finalPrice = None
        itemPrice = None
        quantityPurchased = None
        rewardsReceiptStatus = rcp.get('rewardsReceiptStatus', None)
        totalSpent = rcp.get('totalSpent', None)
        userId = rcp.get('userId', None)        

        rcp_id_list.append(rcp_id)
        bonusPointsEarned_list.append(bonusPointsEarned)
        createDate_list.append(createDate)
        dateScanned_list.append(dateScanned)
        finishedDate_list.append(finishedDate)
        modifyDate_list.append(modifyDate)
        pointsAwardedDate_list.append(pointsAwardedDate)
        pointsEarned_list.append(pointsEarned)
        purchaseDate_list.append(purchaseDate)
        purchasedItemCount_list.append(purchasedItemCount)
        barcode_list.append(barcode)
        description_list.append(description)
        finalPrice_list.append(finalPrice)
        itemPrice_list.append(itemPrice)
        quantityPurchased_list.append(quantityPurchased)
        rewardsReceiptStatus_list.append(rewardsReceiptStatus)
        totalSpent_list.append(totalSpent)
        userId_list.append(userId)   
        
receipts_df = pd.DataFrame({'rcp_id': rcp_id_list})

column_nm ={'bonusPointsEarned':bonusPointsEarned_list, 'createDate':createDate_list, 
            'dateScanned':dateScanned_list, 'finishedDate':finishedDate_list,
            'modifyDate':modifyDate_list, 'pointsAwardedDate':pointsAwardedDate_list, 
            'pointsEarned':pointsEarned_list, 'purchaseDate':purchaseDate_list, 
            'purchasedItemCount':purchasedItemCount_list, 'barcode':barcode_list, 
            'description':description_list, 'finalPrice':finalPrice_list, 'itemPrice':itemPrice_list,
            'quantityPurchased':quantityPurchased_list, 'rewardsReceiptStatus':rewardsReceiptStatus_list, 
            'totalSpent':totalSpent_list, 'userId':userId_list}


for col_nm, col in column_nm.items():
    receipts_df[col_nm] = col

In [81]:
receipts_df

rcp_id  bonusPointsEarned  createDate dateScanned  \
0     5ff1e1eb0a720f0523000575              500.0  2021-01-03  2021-01-03   
1     5ff1e1bb0a720f052300056b              150.0  2021-01-03  2021-01-03   
2     5ff1e1bb0a720f052300056b              150.0  2021-01-03  2021-01-03   
3     5ff1e1f10a720f052300057a                5.0  2021-01-03  2021-01-03   
4     5ff1e1ee0a7214ada100056f                5.0  2021-01-03  2021-01-03   
...                        ...                ...         ...         ...   
7376  603d0b710a720fde1000042a                NaN  2021-03-01  2021-03-01   
7377  603cf5290a720fde10000413                NaN  2021-03-01  2021-03-01   
7378  603ce7100a7217c72c000405               25.0  2021-03-01  2021-03-01   
7379  603ce7100a7217c72c000405               25.0  2021-03-01  2021-03-01   
7380  603c4fea0a7217c72c000389                NaN  2021-02-28  2021-02-28   

     finishedDate  modifyDate pointsAwardedDate pointsEarned purchaseDate  \
0      2021-01-03  2021-01-03        2021-01-03        500.0   2021-01-02   
1      2021-01-03  2021-01-03        2021-01-03        150.0   2021-01-02   
2      2021-01-03  2021-01-03        2021-01-03        150.0   2021-01-02   
3            None  2021-01-03              None            5   2021-01-02   
4      2021-01-03  2021-01-03        2021-01-03          5.0   2021-01-02   
...           ...         ...               ...          ...          ...   
7376         None  2021-03-01              None         None         None   
7377         None  2021-03-01              None         None         None   
7378         None  2021-03-01              None         25.0   2020-08-16   
7379         None  2021-03-01              None         25.0   2020-08-16   
7380         None  2021-02-28              None         None         None   

      purchasedItemCount       barcode  \
0                    5.0          4011   
1                    2.0          4011   
2                    2.0  028400642255   
3                    1.0          None   
4                    4.0          4011   
...                  ...           ...   
7376                 NaN          None   
7377                 NaN          None   
7378                 2.0    B076FJ92M4   
7379                 2.0    B07BRRLSVC   
7380                 NaN          None   

                                            description finalPrice itemPrice  \
0                                        ITEM NOT FOUND      26.00     26.00   
1                                        ITEM NOT FOUND          1         1   
2     DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00     10.00   
3                                                  None       None      None   
4                                        ITEM NOT FOUND      28.00     28.00   
...                                                 ...        ...       ...   
7376                                               None       None      None   
7377                                               None       None      None   
7378  mueller austria hypergrind precision electric ...      22.97     22.97   
7379  thindust summer face mask - sun protection nec...      11.99     11.99   
7380                                               None       None      None   

      quantityPurchased rewardsReceiptStatus totalSpent  \
0                   5.0             FINISHED      26.00   
1                   1.0             FINISHED      11.00   
2                   1.0             FINISHED      11.00   
3                   NaN             REJECTED      10.00   
4                   4.0             FINISHED      28.00   
...                 ...                  ...        ...   
7376                NaN            SUBMITTED       None   
7377                NaN            SUBMITTED       None   
7378                1.0             REJECTED      34.96   
7379                1.0             REJECTED      34.96   
7380                NaN            SUBMITTED       None   


In [82]:
receipts_df['rewardsReceiptStatus']

0        FINISHED
1        FINISHED
2        FINISHED
3        REJECTED
4        FINISHED
          ...    
7376    SUBMITTED
7377    SUBMITTED
7378     REJECTED
7379     REJECTED
7380    SUBMITTED
Name: rewardsReceiptStatus, Length: 7381, dtype: object

In [108]:
receipts_df['barcode'].unique()

array(['4011', '028400642255', None, '1234', '046000832517',
       '013562300631', '034100573065', '075925306254', '041000168468',
       '2700719497082', '079400066619', '051500720011', '087684001127',
       '071040063102', '029000079236', '041129002292', '043000946060',
       '044700009888', '021000012961', '021000678358', '044700033302',
       '021000059232', '043000004944', '759283400082', '025800000135',
       '043000043486', '029000024748', '044700002810', '043000079904',
       '044000000745', '046704085905', '021000057832', '021000002917',
       '044700019917', '043000035818', '070085046583', '044700073377',
       '021000055920', '021000051885', '022174070214', '043000012871',
       '043000077467', '043000008836', '021000068364', '021000013869',
       '044700030479', '041258081601', '043000075821', '021000667543',
       '013120013966', '021000068760', '070277290107', '041129088005',
       '013000993906', '043000065198', '021000050765', '021000032778',
       '0430000

#### rewardsReceiptStatus is rejected but pointsAwardedDate is not None

In [90]:
receipts_df[(receipts_df['rewardsReceiptStatus'] == 'REJECTED') & (receipts_df['pointsAwardedDate'].isna() == False)].drop_duplicates()

rcp_id  bonusPointsEarned  createDate dateScanned  \
24    5f9c74f70a7214ad07000037              750.0  2020-10-30  2020-10-30   
25    5f9c74f70a7214ad07000037              750.0  2020-10-30  2020-10-30   
6116  5fa8d5730a7214adc30001c3              750.0  2020-11-08  2020-11-08   
6117  5fa8d5730a7214adc30001c3              750.0  2020-11-08  2020-11-08   
6854  5fb1bd510a7214ad890000e3              750.0  2020-11-15  2020-11-15   
6855  5fb1bd510a7214ad890000e3              750.0  2020-11-15  2020-11-15   
6883  5fb2da580a7214ad89000114              750.0  2020-11-16  2020-11-16   
6884  5fb2da580a7214ad89000114              750.0  2020-11-16  2020-11-16   

     finishedDate  modifyDate pointsAwardedDate pointsEarned purchaseDate  \
24           None  2021-01-03        2020-10-30        750.0   2020-10-29   
25           None  2021-01-03        2020-10-30        750.0   2020-10-29   
6116         None  2021-01-26        2020-11-08        750.0   2020-11-07   
6117         None  2021-01-26        2020-11-08        750.0   2020-11-07   
6854         None  2021-02-01        2020-11-15        750.0   2020-11-14   
6855         None  2021-02-01        2020-11-15        750.0   2020-11-14   
6883         None  2021-02-01        2020-11-16        750.0   2020-11-15   
6884         None  2021-02-01        2020-11-16        750.0   2020-11-15   

      purchasedItemCount       barcode description finalPrice itemPrice  \
24                  11.0  075925306254        None          1         1   
25                  11.0          None        None       None      None   
6116                11.0  075925306254        None          1         1   
6117                11.0          None        None       None      None   
6854                11.0  075925306254        None          1         1   
6855                11.0          None        None       None      None   
6883                11.0  075925306254        None          1         1   
6884                11.0          None        None       None      None   

      quantityPurchased rewardsReceiptStatus totalSpent  \
24                  1.0             REJECTED       1.00   
25                  NaN             REJECTED       1.00   
6116                1.0             REJECTED       1.00   
6117                NaN             REJECTED       1.00   
6854                1.0             REJECTED       1.00   
6855                NaN             REJECTED       1.00   
6883                1.0             REJECTED       1.00   
6884                NaN             REJECTED       1.00   

                        userId  
24    5f9c74f7c88c1415cbddb839  
25    5f9c74f7c88c1415cbddb839  
6116  5fa8d573770b516dda4f4d21  
6117  5fa8d573770b516dda4f4d21  
6854  5fb1bd519c09d811e5211cef  
6855  5fb1bd519c09d811e5211cef  
6883  5fb2da582357ec11d4ecc403  
6884  5fb2da582357ec11d4ecc403

#### For each recript id, check if the total quantityPurchased is eauql to purchasedItemCount

In [106]:
sum_purchase = pd.DataFrame(receipts_df.groupby('rcp_id')['quantityPurchased'].apply('sum')).reset_index()
total_purchase = receipts_df[['rcp_id', 'purchasedItemCount']]
merge = total_purchase.merge(sum_purchase, on = 'rcp_id', how = 'left')
merge[merge['purchasedItemCount'] != merge['quantityPurchased']]['rcp_id'].nunique()

528

#### Conclusion for data quality issues
1. One data quality issue I find in the data is there is exist some rows with conflict information. For example, the table above shows there are 4 rows rcp_id where the  rewardsReceiptStatus is 'REJECTED' but pointsAwardedDate is not None.

2. The second issue I find the there are 528 receipts where the sum of the quantityPurchased of all th items under that receipt is not equal to the purchasedItemCount.